### PR TITLE
[#11586] Remove trimming of gmail.com for user accounts

### PR DESCRIPTION
**Pull request by xyliew25**
Part of #11586

**Outline of Solution**
1. Remove trimming of `@gmail.com` when saving `googleId` in the `Account`, `Instructor`, `CourseStudent` and `StudentProfile` entities and query for both `googleId` and `gmail`, i.e. `googleId` appended with `@gmail.com`, when searching for users in the database.
2. Write and run a data migration script to set `googleId` to its corresponding `gmail` if it is not already.
3. Remove querying of `googleId` when searching for users in the database.

Maybe rename `googleId` to `userId`?

### CODE REVIEW

File: src/client/java/teammates/client/scripts/DataMigrationForGoogleIdToGmail.java
Comment (wkurniawan07) [inline src/client/java/teammates/client/scripts/DataMigrationForGoogleIdToGmail.java:28]: This is not the right condition. A Google ID can end with some other domain. It's only when there is no domain that you need to append `@gmail.com` to it.

**Review by wkurniawan07 — CHANGES_REQUESTED** [review]
Aware that it's still draft, but there are some quite glaring mistakes

File: src/client/java/teammates/client/scripts/DataMigrationForGoogleIdToGmail.java
Comment (xyliew25) [reply src/client/java/teammates/client/scripts/DataMigrationForGoogleIdToGmail.java]: Hm, may I know why is it that a Google ID can end with other domains? Is this a legacy thing?

Comment (wkurniawan07) [reply src/client/java/teammates/client/scripts/DataMigrationForGoogleIdToGmail.java]: That's not legacy; that's how Google OAuth2 works.

Comment (xyliew25) [reply src/client/java/teammates/client/scripts/DataMigrationForGoogleIdToGmail.java]: Ah I see, I didn't know that one can create a Google account using a non-Gmail email address.

File: src/main/java/teammates/storage/api/InstructorsDb.java
Comment (wkurniawan07) [inline src/main/java/teammates/storage/api/InstructorsDb.java:399]: The split should be on `@gmail.com`, not just `@`

File: src/main/java/teammates/storage/api/InstructorsDb.java
Comment (wkurniawan07) [inline src/main/java/teammates/storage/api/InstructorsDb.java:402]: And here it should take place only if the ID/email has no `@` in it

**Review by wkurniawan07 — CHANGES_REQUESTED** [review]
The split rule needs to be fixed, otherwise implementation is ok

File: src/main/java/teammates/storage/api/InstructorsDb.java
Comment (xyliew25) [reply src/main/java/teammates/storage/api/InstructorsDb.java:399]: To confirm, say if the `googleIdOrEmail` is a gmail, e.g. `abc@gmail.com`, both `abc` and `abc@gmail.com` should be queried; if the `googleIdOrEmail` is a non-gmail, e.g. `test@example.com`, we can just query for `test@example.com`, right?

Comment (wkurniawan07) [reply src/main/java/teammates/storage/api/InstructorsDb.java:399]: Yes. The latter has no ambiguity, so to speak.

File: src/main/java/teammates/storage/api/InstructorsDb.java
Comment (xyliew25) [reply src/main/java/teammates/storage/api/InstructorsDb.java]: I assume you're referring to the else block. My mistake previously, there will be no such cases, right?

Comment (wkurniawan07) [reply src/main/java/teammates/storage/api/InstructorsDb.java]: There will be for a short while after the release, because existing logged in users will still have their cookies in the form of without `@gmail.com`. So this condition still needs to be added.

Comment (xyliew25) [reply src/main/java/teammates/storage/api/InstructorsDb.java]: Fyi, the current query logic is such that if the `googleId` is a gmail, e.g. `abc@gmail.com`, then both `abc` and `abc@gmail.com` will be queried; else the `googleId` is queried as it is. Not sure if I get your concern right, but the above case will be covered as well, i.e. by the latter case.

Comment (wkurniawan07) [reply src/main/java/teammates/storage/api/InstructorsDb.java]: Unless we plan to run the script only one week after the new version is released, you need to cover this. Consider:
- Day 1, system is still at V8.18.0, someone logs in and recognized as `x` (no `@gmail.com`).
- V8.19.0 (or whichever version has this fix) is released in day 2.
- The migration script is run on day 3. `x` is now `x@gmail.com` in the DB.
- The same person logs in back on day 4, with the same login cookie that s/he gets from day 1 (remember that cookies are good for 7 days). S/he will not get the entities with `x` as the ID as they have been migrated.

While it's not a big deal (s/he can just logout and re-login), it will be good if we can cover this even if temporarily.

Comment (xyliew25) [reply src/main/java/teammates/storage/api/InstructorsDb.java]: I see, I overlooked the period between after migration and merging of new code.

**Review by wkurniawan07 — APPROVED** [review]

**Comment by damithc** [comment]
@xyliew25 after this code was deployed, there was an error of this nature 

>java.lang.AssertionError: Trying to create a course for a non-existent instructor abcd@gmail.com
        at teammates.logic.core.CoursesLogic.createCourseAndInstructor(CoursesLogic.java:104)

Is the above related to this PR? The user is shown as `abcd` (not `abcd@gmail.com`) when I search his email.
There was only one such report so far, but I switch back to v8.18.0 for the time being just in case.

**Comment by damithc** [comment]
> There was only one such report so far, but I switch back to v8.18.0 for the time being just in case.

The user says the problem went away after I reverted to v8.18.0
While not conclusive, this increases the probability that the issue was related to something in v8.19.0

**Comment by xyliew25** [comment]
@damithc indeed, the error was caused by this PR as there was some missing code to deal with the old data. Very sorry for my oversight and the inconvenience caused! Will submit a PR soon.

**Comment by damithc** [comment]
> @damithc indeed, the error was caused by this PR as there was some missing code to deal with the old data. Very sorry for my oversight and the inconvenience caused! Will submit a PR soon.

No worries. These things happen when implementing major changes, especially if involving legacy data.

**Comment by damithc** [comment]
@xyliew25 another user (a student, this time) reported a similar issue w.r.t. response submission

image

The user can access her home page and see the list of courses but when she tries to submit responses, she gets the above error.

Again, there was only one such report so far (because most students are not registered), and I have reverted to the previous version for the time being. Could you check if there is indeed a problem?

**Comment by xyliew25** [comment]
@damithc will do

**Comment by damithc** [comment]
@xyliew25 while you are at it, check admin side too. I realized I'll have to update `build.properties` as follows:

previous:
>app.admins=abc

new:
>app.admins=abc@gmail.com

Not sure the above will cause problems when I'm using the UI to access admin features.

**Comment by wkurniawan07** [comment]
The update of `build.properties` values for admin is expected. In fact, the update should apply to all maintainers as well.

**Comment by damithc** [comment]
> In fact, the update should apply to all maintainers as well.

Right, will update that too.

**Comment by damithc** [comment]
@xyliew25 I deployed v8.19.2 but I can't access admin features using that version although I already updated `build.properties` as `app.admins=abc@gmail.com`

I reverted back to v8.18 for the time being, while you investigate this issue. Also check if maintainers can face a similar issue when they access 'maintainers-only' features.

**Comment by wkurniawan07** [comment]
@damithc that's probably because you're logged in still as `abc`. Remember that login cookies apply for 7 days.
If anything, you may want to just include both variants of `abc` and `abc@gmail.com`, at least for the time being.

**Comment by damithc** [comment]
> @damithc that's probably because you're logged in still as `abc`. Remember that login cookies apply for 7 days.
> If anything, you may want to just include both variants of `abc` and `abc@gmail.com`, at least for the time being.

Got it. Let me try that and report here shortly.

**Comment by damithc** [comment]
> Got it. Let me try that and report here shortly.

A simple logout/login worked. So, no issue on using admin features. I've switched the live server to v8.19.2 again.

**Pull request merged**
